lib/vtls/mbedtls.c

```diff
@@ -686,6 +686,13 @@ mbed_connect_step2(struct Curl_easy *data, struct connectdata *conn,
     if(!p)
       return CURLE_OUT_OF_MEMORY;
 
+    pubkey = malloc(PUB_DER_MAX_BYTES);
```

Comment:
Is `PUB_DER_MAX_BYTES` always needed or can it be smaller and grown when needed?

Reply:
So I analyzed this question more. It seems possible, but it will not be easy.
From the mbedtls code you can deduce what size buffer is needed.

For RSA:
```
/ *
 * RSA public keys:
 * SubjectPublicKeyInfo :: = SEQUENCE {1 + 3
 * algorithm AlgorithmIdentifier, 1 + 1 (sequence)
 * + 1 + 1 + 9 (rsa oid)
 * + 1 + 1 (params null)
 * subjectPublicKey BIT STRING} 1 + 3 + (1 + below)
 * RSAPublicKey :: = SEQUENCE {1 + 3
 * modulus INTEGER, - n 1 + 3 + MPI_MAX + 1
 * publicExponent INTEGER - e 1 + 3 + MPI_MAX + 1
 *}
 * /
```
For EC:
```
/ *
 * EC public keys:
 * SubjectPublicKeyInfo :: = SEQUENCE {1 + 2
 * algorithm AlgorithmIdentifier, 1 + 1 (sequence)
 * + 1 + 1 + 7 (ec oid)
 * + 1 + 1 + 9 (namedCurve oid)
 * subjectPublicKey BIT STRING 1 + 2 + 1 [1]
 * + 1 (point format) [1]
 * + 2 * ECP_MAX (coords) [1]
 *}
 * /
```

The problem for RSA is determining the size of the modulus parameter and the publicExponent (instead of taking the value of MPI_MAX). For the EC, the problem is to determine the size of the Q parameter (instead of taking the value ECP_MAX).

If the public key is RSA, you can get the N (modulus) and E (exponent) parameters using the mbedtls_rsa_import function, read their size by the mbedtls_mpi_size function, sum the values ​​and add an additional 38 bytes.

Similarly for EC: the mbedtls_pk_ec function returns the mbedtls_ecp_keypair structure that contains the Q parameter. Then read the X and Y coordinates, sum and add an additional 30 bytes.

Disadvantages:
- adding new asymmetric keys will require changes in curl as well.
- RSA and EC key parameters have been marked as private in the mbedtls 3.0.0 version - it is uncertain whether they will not be removed in the future for the appropriate functions in the API.

The best solution would of course be to write such a function on the mbedtls side. However, I have not found any premises for the mbedtls team to implement such a function.

Reply:
Thanks for the exhaustive answer! With this in mind I agree that just allocating the known upper limit is clearly the best option.